File: crates/kuben-platform/src/controller/project.rs
//! Project status: number of live environments + `Ready`.

use std::{sync::Arc, time::Duration};

use futures::StreamExt;
use kube::{
    Api, ResourceExt,
    api::{ListParams, Patch, PatchParams},
    runtime::{Controller, controller::Action, reflector::ObjectRef, watcher},
};
use kuben_crd::{Environment, FIELD_MANAGER, Project, ProjectStatus, condition::READY};
use serde_json::json;
use tokio_util::sync::CancellationToken;

use super::{Ctx, Result, condition, error_policy};

pub async fn run(ctx: Arc<Ctx>, token: CancellationToken) -> anyhow::Result<()> {
    let client = ctx.client.clone();
    Controller::new(Api::<Project>::all(client.clone()), watcher::Config::default())
        .watches(
            Api::<Environment>::all(client),
            watcher::Config::default(),
            |env| Some(ObjectRef::<Project>::new(&env.spec.project)),
        )
        .graceful_shutdown_on(token.cancelled_owned())
        .run(reconcile, error_policy, ctx)
        .for_each(|_| std::future::ready(()))
        .await;
    Ok(())
}

#[allow(clippy::needless_pass_by_value)] // signature required by `Controller::run`
async fn reconcile(project: Arc<Project>, ctx: Arc<Ctx>) -> Result<Action> {
    let name = project.name_any();
    // Projects are few and environments per project fewer; a LIST is cheaper
    // than keeping a second cache in sync.
    let envs = Api::<Environment>::all(ctx.client.clone())
        .list(&ListParams::default())
        .await?;
    let count = envs
        .items
        .iter()
        .filter(|e| e.spec.project == name && e.metadata.deletion_timestamp.is_none())
        .count();
    let previous = project
        .status
        .as_ref()
        .map_or(&[][..], |s| s.conditions.as_slice());
    let status = ProjectStatus {
        observed_generation: project.metadata.generation,
        environments: u32::try_from(count).unwrap_or(u32::MAX),
        conditions: vec![condition(
            previous,
            READY,
            true,
            "Reconciled",
            "",
            project.metadata.generation,
        )],
    };
    let patch = json!({ "apiVersion": "kuben.dev/v1alpha1", "kind": "Project", "status": status });
    Api::<Project>::all(ctx.client.clone())
        .patch_status(
            &name,
            &PatchParams::apply(FIELD_MANAGER).force(),
            &Patch::Apply(&patch),
        )
        .await?;
    ctx.succeeded(project.as_ref());
    Ok(Action::requeue(Duration::from_mins(15)))
}

File: apps/kuben/internal/kube/controller/project.go
package controller

import (
	"context"
	"fmt"
	"math"
	"time"

	"sigs.k8s.io/controller-runtime/pkg/client"
	"sigs.k8s.io/controller-runtime/pkg/reconcile"

	"github.com/Teamtem-dev/kuben/packages/api/v1alpha1"
)

// Project status: the number of live environments, and Ready
// (controller::project).

// projectResync is how often a Project is reconciled without a change.
const projectResync = 15 * time.Minute

type projectReconciler struct{ *shared }

// Reconcile writes the Project's status.
func (r projectReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
	var project v1alpha1.Project
	if err := r.client.Get(ctx, req.NamespacedName, &project); err != nil {
		return reconcile.Result{}, client.IgnoreNotFound(err)
	}
	// Projects are few and environments per project fewer; a LIST is
	// cheaper than keeping a second cache in sync.
	var envs v1alpha1.EnvironmentList
	if err := r.reader.List(ctx, &envs); err != nil {
		return reconcile.Result{}, fmt.Errorf("listing environments: %w", err)
	}
	var count uint64
	for i := range envs.Items {
		e := &envs.Items[i]
		if e.Spec.Project == project.Name && e.DeletionTimestamp == nil {
			count++
		}
	}
	var previous []v1alpha1.Condition
	if project.Status != nil {
		previous = project.Status.Conditions
	}
	generation := generationOf(&project.ObjectMeta)
	status := v1alpha1.ProjectStatus{
		ObservedGeneration: generation.Ptr(),
		Environments:       uint32(min(count, math.MaxUint32)),
		Conditions: v1alpha1.Conditions{
			Condition(r.clock, previous, v1alpha1.ConditionReady, true, "Reconciled", "", generation),
		},
	}
	if err := applyStatus(ctx, r.client, v1alpha1.ProjectKind, "", project.Name, status); err != nil {
		return reconcile.Result{}, err
	}
	return reconcile.Result{RequeueAfter: projectResync}, nil
}

// projectOfEnvironment maps an Environment to its Project.
func projectOfEnvironment(_ context.Context, env *v1alpha1.Environment) []reconcile.Request {
	return []reconcile.Request{{NamespacedName: client.ObjectKey{Name: env.Spec.Project}}}
}
